iot_packet_verifier/src/burner.rs

```diff
@@ -0,0 +1,252 @@
+
+        // Fetch escrow account
+        let ddc_key = pdas::delegated_data_credits(&self.program_cache.sub_dao, &payer);
+        let account_data = self.provider.get_account_data(&ddc_key).await?;
+        let mut account_data = account_data.as_ref();
+        let escrow_account =
+            DelegatedDataCreditsV0::try_deserialize(&mut account_data)?.escrow_account;
+
+        tracing::info!("Burning {} DC from {}", amount, payer);
+
+        let instructions = {
+            let request = RequestBuilder::from(
+                data_credits::id(),
+                &self.cluster,
+                std::rc::Rc::new(Keypair::from_bytes(&self.keypair).unwrap()),
+                Some(CommitmentConfig::confirmed()),
+                RequestNamespace::Global,
+            );
+
+            let accounts = accounts::BurnDelegatedDataCreditsV0 {
+                sub_dao_epoch_info,
+                dao: self.program_cache.dao,
+                sub_dao: self.program_cache.sub_dao,
+                account_payer: self.program_cache.account_payer,
+                data_credits: self.program_cache.data_credits,
+                delegated_data_credits: pdas::delegated_data_credits(
+                    &self.program_cache.sub_dao,
+                    &payer,
+                ),
+                token_program: spl_token::id(),
+                helium_sub_daos_program: helium_sub_daos::id(),
+                system_program: solana_program::system_program::id(),
+                dc_burn_authority: self.program_cache.dc_burn_authority,
+                dc_mint: self.program_cache.dc_mint,
+                escrow_account,
+                registrar: self.program_cache.registrar,
+            };
+            let args = instruction::BurnDelegatedDataCreditsV0 {
+                args: data_credits::BurnDelegatedDataCreditsArgsV0 {
+                    amount: amount as u64,
+                },
+            };
+
+            // As far as I can tell, the instructions function does not actually have any
+            // error paths.
+            request
+                .accounts(accounts)
+                .args(args)
+                .instructions()
+                .unwrap()
+        };
+
+        let blockhash = self.provider.get_latest_blockhash().await?;
+        let signer = Keypair::from_bytes(&self.keypair).unwrap();
+
+        let tx = Transaction::new_signed_with_payer(
+            &instructions,
+            Some(&signer.pubkey()),
+            &[&signer],
+            blockhash,
+        );
+
+        let signature = self.provider.send_and_confirm_transaction(&tx).await?;
+        tracing::info!(
+            "Successfully burned data credits. Transaction: {}",
+            signature
+        );
+
+        // Now that we have successfully executed the burn and are no long in
+        // sync land, we can remove the amount burned.
+        sqlx::query(
+            r#"
+            UPDATE pending_burns SET
+              amount = amount - $1,
```

Comment:
in theory this should just be $0 right, since you got the `amount` value from a query at the beginning of this function? if there's any chance this value could be updated out-of-band from this burn function should we enforce some guard that ensures that `amount` in the DB never goes into the negative?

Reply:
It _may_ be zero or greater than. It is possible for the verification process to add burn amounts during this - or at least ideally it can. No other process can subtract from the burn amount besides this one though, so it will never become negative

Reply:
👍